### PR TITLE
Add documentation for running visualisation snapshot tests

### PR DESCRIPTION
**Pull request by javabster**
### Summary
A number of community contributors are looking at tackling issues relating to visualisations. This requires them to update the visualisation snapshot tests. This PR adds documentation to the contributing.md explaining how to run snapshot tests and update them.

### CODE REVIEW

**Comment by jsistos** [comment]
As a side note, every time ```test/ipynb/mpl_tester.ipynb``` is used to run the snapshot tests, some of its metadata changes. While the file can be easily taken out of a commit, maybe it would be useful to include it in the ```.gitignore``` file.

**Comment by 1ucian0** [comment]
> As a side note, every time `test/ipynb/mpl_tester.ipynb` is used to run the snapshot tests, some of its metadata changes. While the file can be easily taken out of a commit, maybe it would be useful to include it in the `.gitignore` file.

I use https://pypi.org/project/nb-clean/ for "cleaning" notebooks before pushing them. Maybe add this as a recommendation in the doc?

Other possibility is to add nb-clean as a pre-commit hook. But this means yet another req-dev. Maybe @mtreinish has an opinion.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -312,6 +312,41 @@ Alternatively, the `make test_ci` target can be used instead of
 `make test` in order to run in a setup that replicates the configuration
 we used in our CI systems more closely.
 
+### Snapshot Testing for Visualizations
+
+If you are working on code that makes changes to any matplotlib visualisations
```
Comment (1ucian0) [inline CONTRIBUTING.md:317]: our plan is to add circuit latex drawer to this schema at some point. So I think "matplotlib" is a bit too narrow. But "visualization" is a bit too wide (for example, circuit text drawer is a visualization that wont be not included).
 
```suggestion
If you are working on code that makes changes to rich visualisations 
```

Is that too obscure?

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -312,6 +312,41 @@ Alternatively, the `make test_ci` target can be used instead of
 `make test` in order to run in a setup that replicates the configuration
 we used in our CI systems more closely.
 
+### Snapshot Testing for Visualizations
+
+If you are working on code that makes changes to any matplotlib visualisations 
+you will need to check that your changes don't break any snapshot tests, and add 
+new tests where necessary. You can do this as follows:
+
+1. Make sure you have pushed your branch with the latest changes.
```
Comment (1ucian0) [inline CONTRIBUTING.md:321]:
```suggestion
1. Make sure you have pushed your branch with the latest changes into your remote.
```

File: CONTRIBUTING.md (hunk 1 above)
Comment (1ucian0) [reply CONTRIBUTING.md:317]: The rest of the explanation is very much focus on circuit matplotlib visualization. So probably is better as it is and then come back here to update as you tests are being added.

```suggestion
If you are working on code that makes changes to the circuit matplotlib drawer 
```

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -312,6 +312,41 @@ Alternatively, the `make test_ci` target can be used instead of
...
+2. Go to link: `https://mybinder.org/v2/gh/<github_user>/qiskit-terra/<branch>?urlpath=apps/test/ipynb/mpl_tester.ipynb`.
```
Comment (1ucian0) [inline CONTRIBUTING.md:322]:
```suggestion
2. Go to link: `https://mybinder.org/v2/gh/<github_user>/<repo>/<branch>?urlpath=apps/test/ipynb/mpl_tester.ipynb`. For example, if your GitHub username is `username`, your forked repo has the same name the original, and your branch is `my_awesome_new_feature`, you should visit https://mybinder.org/v2/gh/username/qiskit-terra/my_awesome_new_feature?urlpath=apps/test/ipynb/mpl_tester.ipynb`.
```

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -312,6 +312,41 @@ Alternatively, the `make test_ci` target can be used instead of
...
 
+### Snapshot Testing for Visualizations
+
+If you are working on code that makes changes to any matplotlib visualisations 
+you will need to check that your changes don't break any snapshot tests, and add 
+new tests where necessary. You can do this as follows:
+
+1. Make sure you have pushed your branch with the latest changes.
+2. Go to link: `https://mybinder.org/v2/gh/<github_user>/qiskit-terra/<branch>?urlpath=apps/test/ipynb/mpl_tester.ipynb`. 
+This opens a jupyter application running in the cloud that automatically runs
```
Comment (1ucian0) [inline CONTRIBUTING.md:323]:
```suggestion
This opens a Jupyter Notebook application running in the cloud that automatically runs 
```

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -312,6 +312,41 @@ Alternatively, the `make test_ci` target can be used instead of
...
+you will need to check that your changes don't break any snapshot tests, and add 
+new tests where necessary. You can do this as follows:
+
+1. Make sure you have pushed your branch with the latest changes.
+2. Go to link: `https://mybinder.org/v2/gh/<github_user>/qiskit-terra/<branch>?urlpath=apps/test/ipynb/mpl_tester.ipynb`. 
+This opens a jupyter application running in the cloud that automatically runs 
+the snapshot tests (note this may take some time to finish loading).
+3. If any tests fail you should see sets of 3 images for each failed test 
+(left: reference image, middle: your test result, right: differences):
+   <img width="995" alt="Screenshot_2021-03-26_at_14 13 54" src="https://user-images.githubusercontent.com/23662430/112663508-d363e800-8e50-11eb-9478-6d665d0ff086.png">
```
Comment (1ucian0) [inline CONTRIBUTING.md:327]:
```suggestion
3. 3. Each test should be a set of 3 images  (left: reference image, middle: your test result, right: differences). Passed tests are folded. If a test fails, you will see a situation like this one:
   <img width="995" alt="Screenshot_2021-03-26_at_14 13 54" src="https://user-images.githubusercontent.com/23662430/112663508-d363e800-8e50-11eb-9478-6d665d0ff086.png">
```

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -312,6 +312,41 @@ Alternatively, the `make test_ci` target can be used instead of
...
+4. Fix any broken tests. Working on code for one aspect of the visualisations 
+can sometimes result in minor changes elsewhere to spacing etc. In these cases 
+you just need to update the reference images as follows:
+    - download the mismatched images (link at top of jupyter notebook output)
+    - unzip the folder
+    - copy and paste the new images into `qiskit-terra/test/ipynb/mpl/references`, 
+  replacing the existing reference images
+    - commit and push your changes, restart the jupyter app in your browser. The
```
Comment (1ucian0) [inline CONTRIBUTING.md:335]:
```suggestion
    - add, commit, and push your changes, restart the Jupyter Notebook app in your browser. The 
```

Comment (1ucian0) [reply CONTRIBUTING.md:335]: Not fully sure if the binder env will update the branch. You might need to create a new env by visiting the link in step 2. This needs to be confirmed.

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -312,6 +312,41 @@ Alternatively, the `make test_ci` target can be used instead of
...
+4. Fix any broken tests. Working on code for one aspect of the visualisations 
+can sometimes result in minor changes elsewhere to spacing etc. In these cases 
+you just need to update the reference images as follows:
+    - download the mismatched images (link at top of jupyter notebook output)
+    - unzip the folder
+    - copy and paste the new images into `qiskit-terra/test/ipynb/mpl/references`, 
+  replacing the existing reference images
+    - commit and push your changes, restart the jupyter app in your browser. The 
+  tests should now pass.
+5. Add new snapshot tests covering your new contributions.
```
Comment (1ucian0) [inline CONTRIBUTING.md:337]:
```suggestion
5. Add new snapshot tests covering your new features or extensions.
```

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -312,6 +312,41 @@ Alternatively, the `make test_ci` target can be used instead of
...
+    , where you can also find existing tests to use as a guide.
+    - commit and push your changes, restart the jupyter app in your browser. 
+    As this is the first time you run your new tests there won't be any reference 
+    images to compare to. Instead you should see an option in the list of tests 
+    to download the new images, like so:
+
+    <img width="1002" alt="Screenshot_2021-03-26_at_15 38 31" src="https://user-images.githubusercontent.com/23662430/112665215-b9c3a000-8e52-11eb-89e7-b18550718522.png">
+
+    - download the new images, then copy and paste into `qiskit-terra/test/ipynb/mpl/references`
+    - commit and push your changes, restart the jupyter app in your browser. The
```
Comment (1ucian0) [inline CONTRIBUTING.md:348]:
```suggestion
    - add, commit, and push your changes, restart the Jupyter Notebook app in your browser. The 
```

Comment (1ucian0) [reply CONTRIBUTING.md:348]: Same as https://github.com/Qiskit/qiskit-terra/pull/6088/files#r607731246

**Comment by javabster** [comment]
thanks for the review @1ucian0 😄 I've made the changes you suggested (just tweaked the grammar a bit). Ready for re-review when you get a moment

**Review by 1ucian0 — APPROVED** [review]
Thanks Abby!

**Pull request merged**